### PR TITLE
OLD: [Bing session caching]

### LINKED ISSUE #183: Cache Bing Maps sessions within the process
Within a single process, we should avoid repeatedly starting new Bing Maps sessions for the same layer. Instead, we should cache the session per key/layer.

### PR DESCRIPTION
**Pull request by nithinp7**
Caches Bing session keys for the duration of the application.

Fixes #183

### CODE REVIEW

**Comment by nithinp7** [comment]
@kring Can you have a look? Reading through the related code made me wonder if the session key is already getting cached and reused through the regular caching system 😅. The asset accessor used in `IonRasterOverlay::createTileProvider` (which is where I assumed the new session key is requested) is a `CachingAssetAccessor` that trickles down from the `TilesetExternals` of the owning tileset. Can you let me know if I am missing something in my understanding?

**Comment by nithinp7** [comment]
After offline discussion it looks like the ion response contains a no-cache header when retrieving the Bing session id. The caching system respects this and doesn't cache, so the above question I had seems to be resolved.

**Review by baothientran — COMMENTED** [review]
@nithinp7  This looks good to me. I only have one comment below about data race

**Comment by baothientran** [comment]
Also can you update CHANGE.md

File: Cesium3DTiles/src/IonRasterOverlay.cpp [hunk 1]
```diff
@@ -99,6 +128,8 @@ IonRasterOverlay::createTileProvider(
               std::string culture =
                   JsonHelpers::getStringOrDefault(options, "culture", "");
 
+              cachedBingImageryAssets[ionUrl] = {url, key, mapStyle, culture};
```
[the remark this answers is not in the document]
Comment (baothientran) [reply Cesium3DTiles/src/IonRasterOverlay.cpp:131]: maybe moving it to thenInMainThread is a better option?

**Comment by nithinp7** [comment]
@baothientran Thanks for finding that! I moved it to thenInMainThread, can you look over this again when you get a chance?

**Comment by kring** [comment]
@nithinp7 I don't think this is going to have the desired effect. It'll successfully avoid `/v1/assets/x/endpoint` requests to Cesium ion, but then it will still access the Bing metadata service, which is what Bing counts as a new session. If you look at CesiumJS, there is a cache in _both_ `IonImageryProvider` and in the `BingMapsImageryProvider` that is created by the ion one. If I'm not mistaken, it's actually the second of these that is most important.

In fact, I think caching the ion response is a bad idea. After all, if it were a good idea, ion wouldn't add `cache-control: no-cache` and then we'd get it automatically, right?

You could make the same argument for Bing, except that Bing (unlike ion) uses the request to count usage. If they allowed it to be cached, the counting wouldn't work. Our custom caching here is ok because it's only temporary for the current session; we never cache across editor or game sessions. But that's no way (afaik) to specify that behavior to a browser with cache headers: allow caching, but only within the current page.

Long story short:
1. Don't add custom caching for ion responses. If it's ok to cache an ion response, ion itself can tell us so with cache headers and then we will.
2. Do cache Bing Maps metadata responses, but only for the current session (i.e., never persist them to disk).

Sorry I didn't make this very clear before.

**Comment by kring** [comment]
Fixes #183

**Comment by nithinp7** [comment]
There would be a lot of backtracking in order to fix this here so I just started a new super simple PR here: https://github.com/CesiumGS/cesium-native/pull/237.
